# Циклические структуры данных Часть 1

## Цикл по словарю

В Python 3 для перебора словаря используется метод items():

    world = { "afghanistan":30.55, 
              "albania":2.77,
              "algeria":39.21 }
    
    for key, value in world.items() :
        print(key + " -- " + str(value))

Помните словарь `europe`, который содержал названия некоторых европейских стран в качестве ключей и их столицы как соответствующие значения? Давайте напишем цикл для его итерации!

### Инструкции

Напишите цикл `for`, который проходит через каждую пару ключ:значение в `europe`. На каждой итерации должно выводиться столица `x` - это `y`, где `x` - ключ, а `y` - значение пары.

In [1]:
# Определение словаря
europe = {'spain':'madrid', 'france':'paris', 'germany':'berlin',
          'norway':'oslo', 'italy':'rome', 'poland':'warsaw', 'austria':'vienna' }

# Итерация по Europe
for country, capital in europe.items():
    print("Столица " + country + " - это " + capital)

Столица spain - это madrid
Столица france - это paris
Столица germany - это berlin
Столица norway - это oslo
Столица italy - это rome
Столица poland - это warsaw
Столица austria - это vienna


## Цикл по массиву NumPy

Если у вас есть одномерный массив `NumPy`, перебор всех элементов может быть таким простым:

    for x in my_array :
        ...
        
Если у вас есть двумерный массив NumPy, это более сложно. 2D-массив состоит из нескольких 1D-массивов. Чтобы явно перебрать все отдельные элементы многомерного массива, вам понадобится следующий синтаксис:

    for x in np.nditer(my_array) :
        ...
        
Два массива NumPy, которые вы, возможно, узнаете из вводного курса, доступны в вашей сессии Python: np_height, массив NumPy, содержащий рост игроков Major League Baseball, и np_baseball, 2D-массив NumPy, который содержит и рост (первый столбец), и вес (второй столбец) этих игроков.

### Инструкции

- Импортируйте пакет `numpy` под локальным псевдонимом `np`.
- Напишите цикл `for`, который перебирает все элементы в `np_height` и выводит "x дюймов" для каждого элемента, где `x` - значение в массиве.
- Напишите цикл `for`, который посещает каждый элемент массива `np_baseball` и выводит его.

In [2]:
import pandas as pd
import numpy as np
baseball = pd.read_csv('datasets/MLB(baseball).csv')

np_height = np.array(baseball['Height'])
np_baseball = np.array(baseball[['Height', 'Weight']])

# Цикл по np_height
for height in np_height:
    print(str(height) + " дюймов")

# Цикл по np_baseball
for x in np.nditer(np_baseball):
    print(x)

74 дюймов
74 дюймов
72 дюймов
72 дюймов
73 дюймов
69 дюймов
69 дюймов
71 дюймов
76 дюймов
71 дюймов
73 дюймов
73 дюймов
74 дюймов
74 дюймов
69 дюймов
70 дюймов
73 дюймов
75 дюймов
78 дюймов
79 дюймов
76 дюймов
74 дюймов
76 дюймов
72 дюймов
71 дюймов
75 дюймов
77 дюймов
74 дюймов
73 дюймов
74 дюймов
78 дюймов
73 дюймов
75 дюймов
73 дюймов
75 дюймов
75 дюймов
74 дюймов
69 дюймов
71 дюймов
74 дюймов
73 дюймов
73 дюймов
76 дюймов
74 дюймов
74 дюймов
70 дюймов
72 дюймов
77 дюймов
74 дюймов
70 дюймов
73 дюймов
75 дюймов
76 дюймов
76 дюймов
78 дюймов
74 дюймов
74 дюймов
76 дюймов
77 дюймов
81 дюймов
78 дюймов
75 дюймов
77 дюймов
75 дюймов
76 дюймов
74 дюймов
72 дюймов
72 дюймов
75 дюймов
73 дюймов
73 дюймов
73 дюймов
70 дюймов
70 дюймов
70 дюймов
76 дюймов
68 дюймов
71 дюймов
72 дюймов
75 дюймов
75 дюймов
75 дюймов
75 дюймов
68 дюймов
74 дюймов
78 дюймов
71 дюймов
73 дюймов
76 дюймов
74 дюймов
74 дюймов
79 дюймов
75 дюймов
73 дюймов
76 дюймов
74 дюймов
74 дюймов
73 дюймов
72 дюймов
74 дюймов


## Цикл по DataFrame (1)

Итерация по объекту Pandas DataFrame обычно выполняется с помощью метода `iterrows()`. Используя его в цикле `for`, каждое наблюдение перебирается, и на каждой итерации метка строки и собственно содержимое этой строки становятся доступными:

    for lab, row in brics.iterrows() :
        ...
        
В этом и последующих упражнениях вы будете работать с DataFrame под названием `cars`. Он содержит информацию о количестве автомобилей на человека и о том, едут ли люди направо или налево, для семи стран мира.

### Инструкции

Напишите цикл for, который перебирает строки `cars` и на каждой итерации выполняет два вызова `print()`: один для вывода метки строки и второй для вывода всех содержимого строки.

In [3]:
cars = pd.read_csv('datasets/cars.csv', index_col = 0)

# Итерация по строкам автомобилей
for lab, row in cars.iterrows():
    print(lab)
    print(row)

US
cars_per_cap              809
country         United States
drives_right             True
Name: US, dtype: object
AUS
cars_per_cap          731
country         Australia
drives_right        False
Name: AUS, dtype: object
JAP
cars_per_cap      588
country         Japan
drives_right    False
Name: JAP, dtype: object
IN
cars_per_cap       18
country         India
drives_right    False
Name: IN, dtype: object
RU
cars_per_cap       200
country         Russia
drives_right      True
Name: RU, dtype: object
MOR
cars_per_cap         70
country         Morocco
drives_right       True
Name: MOR, dtype: object
EG
cars_per_cap       45
country         Egypt
drives_right     True
Name: EG, dtype: object


## Цикл по DataFrame (2)

Данные строки, которые генерируются при каждом выполнении `iterrows()`, представляют собой объект Pandas Series. Этот формат не очень удобен для вывода. К счастью, вы можете легко выбирать переменные из объекта Pandas Series, используя квадратные скобки:

    for lab, row in brics.iterrows() :
        print(row['country'])
       

### Инструкции

- Используя итераторы `lab` и `row`, измените код в цикле так, чтобы при первой итерации выводилось `US: 809`, при второй `AUS: 731` и так далее.
- Вывод должен иметь форму "`country: cars_per_cap`". Убедитесь, что выводится именно эта строка (с правильным пробелом).
- Вы можете использовать функцию `str()`, чтобы преобразовать целочисленные данные в строку и вывести их вместе с меткой страны.

In [4]:
# Адаптировать цикл

for lab, row in cars.iterrows() :
    # Напечатать (lab)
    print(lab + ": " + str(row['cars_per_cap']))

US: 809
AUS: 731
JAP: 588
IN: 18
RU: 200
MOR: 70
EG: 45


## Добавление столбца (1)
Добавить `length` (длину) названий стран из DataFrame `brics` в новый столбец:

    for lab, row in brics.iterrows() :
        brics.loc[lab, "name_length"] = len(row["country"])
        
Вы можете сделать нечто подобное с DataFrame `cars`.

### Инструкции

- Используйте цикл for, чтобы добавить новый столбец с именем `COUNTRY`, который содержит верхний регистр названий стран из столбца "`country`". Для этого можно использовать метод `upper()` для строк.
- Чтобы убедиться, что ваш код работает, выведите `cars`. 
- Не делайте отступ для этого кода, чтобы он не был частью цикла `for`.

In [5]:
# Код цикла, добавляющего столбец COUNTRY
for lab, row in cars.iterrows():
    cars.loc[lab, "COUNTRY"] = (row["country"]).upper()

# Вывод cars
print(cars)

     cars_per_cap        country  drives_right        COUNTRY
US            809  United States          True  UNITED STATES
AUS           731      Australia         False      AUSTRALIA
JAP           588          Japan         False          JAPAN
IN             18          India         False          INDIA
RU            200         Russia          True         RUSSIA
MOR            70        Morocco          True        MOROCCO
EG             45          Egypt          True          EGYPT


## Добавление столбца (2)

Использование `iterrows()` для итерации по каждому наблюдению объекта Pandas DataFrame легко понять, но не очень эффективно. На каждой итерации создается новый объект Pandas Series.

Если вы хотите добавить столбец в DataFrame, вызывая функцию для другого столбца, метод `iterrows()` в сочетании с циклом for не является предпочтительным способом. Вместо этого вам следует использовать `apply()`.

Сравните версию с `iterrows()` с версией с `apply()`, чтобы получить тот же результат в DataFrame `brics`:

    for lab, row in brics.iterrows() :
        brics.loc[lab, "name_length"] = len(row["country"])
    
    brics["name_length"] = brics["country"].apply(len)
    
Мы можем сделать нечто подобное для вызова метода `upper()` для каждого имени в столбце `country`. Однако `upper()` - это метод, поэтому нам понадобится немного другой подход:

## Инструкции

- Замените цикл for однострочным кодом, использующим `.apply(str.upper)`. Вызов должен дать тот же результат: в `cars` должен быть добавлен столбец `COUNTRY`, содержащий верхний регистр названий стран.
- Как обычно, выведите `cars`, чтобы увидеть результат вашего труда.

In [6]:
# Использование .apply(str.upper)
# for lab, row in cars.iterrows() :
#     cars.loc[lab, "COUNTRY"] = row["country"].upper()

cars["COUNTRY"] = cars["country"].apply(lambda x: x.upper())
print(cars)

     cars_per_cap        country  drives_right        COUNTRY
US            809  United States          True  UNITED STATES
AUS           731      Australia         False      AUSTRALIA
JAP           588          Japan         False          JAPAN
IN             18          India         False          INDIA
RU            200         Russia          True         RUSSIA
MOR            70        Morocco          True        MOROCCO
EG             45          Egypt          True          EGYPT
